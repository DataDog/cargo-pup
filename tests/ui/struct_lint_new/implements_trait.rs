// This product includes software developed at Datadog (https://www.datadoghq.com/) Copyright 2024 Datadog, Inc.

//@compile-flags: --crate-name test_implements_trait
//@compile-flags: --crate-type lib

// This test verifies that the StructMatch::ImplementsTrait matcher works correctly by
// flagging structs that implement a specific trait.

// Define a test trait
pub trait TestTrait {
    fn test_method(&self) -> bool;
}

// Another trait that shouldn't be matched by our lint
pub trait UnrelatedTrait {
    fn unrelated_method(&self) -> i32;
}

// This struct implements TestTrait and should trigger the lint
pub struct ImplementsTestTrait { //~ ERROR: Struct must match pattern 'Compliant*', found 'ImplementsTestTrait'
    field: i32,
}

impl TestTrait for ImplementsTestTrait {
    fn test_method(&self) -> bool {
        true
    }
}

// This struct implements both traits and should also trigger the lint
pub struct ImplementsMultipleTraits { //~ ERROR: Struct must match pattern 'Compliant*', found 'ImplementsMultipleTraits'
    value: String,
}

impl TestTrait for ImplementsMultipleTraits {
    fn test_method(&self) -> bool {
        false
    }
}

impl UnrelatedTrait for ImplementsMultipleTraits {
    fn unrelated_method(&self) -> i32 {
        42
    }
}

// This struct doesn't implement TestTrait, should not trigger the lint
pub struct DoesNotImplementTestTrait {
    data: Vec<u8>,
}

impl UnrelatedTrait for DoesNotImplementTestTrait {
    fn unrelated_method(&self) -> i32 {
        123
    }
}

// This struct implements TestTrait but has a compliant name, so it shouldn't trigger an error
pub struct CompliantTraitImplementor {
    id: u64,
}

impl TestTrait for CompliantTraitImplementor {
    fn test_method(&self) -> bool {
        true
    }
} 
