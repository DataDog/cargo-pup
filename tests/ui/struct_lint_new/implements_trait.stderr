error: Struct must match pattern 'Compliant*', found 'ImplementsTestTrait'
  --> tests/ui/struct_lint_new/implements_trait.rs:20:1
   |
LL | pub struct ImplementsTestTrait {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Rename this struct to match the pattern 'Compliant*'
   = note: Applied by cargo-pup rule 'implements_trait_test'.
   = note: `#[deny(struct_lint_must_be_named)]` on by default

error: Struct must match pattern 'Compliant*', found 'ImplementsMultipleTraits'
  --> tests/ui/struct_lint_new/implements_trait.rs:31:1
   |
LL | pub struct ImplementsMultipleTraits {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Rename this struct to match the pattern 'Compliant*'
   = note: Applied by cargo-pup rule 'implements_trait_test'.

error: aborting due to 2 previous errors

